src/lib/tls/tls_algos.h

```diff
@@ -117,48 +118,81 @@ enum class Group_Params : uint16_t {
-}
+      constexpr Group_Params(Group_Params_Code code) : m_code(code) {}
 
-constexpr bool is_pure_kyber(const Group_Params group) {
-   return group == Group_Params::KYBER_512_R3 || group == Group_Params::KYBER_768_R3 ||
-          group == Group_Params::KYBER_1024_R3;
-}
+      constexpr Group_Params(uint16_t code) : m_code(static_cast<Group_Params_Code>(code)) {}
 
-constexpr bool is_hybrid(const Group_Params group) {
-   return group == Group_Params::HYBRID_X25519_KYBER_512_R3_CLOUDFLARE ||
-          group == Group_Params::HYBRID_X25519_KYBER_768_R3_CLOUDFLARE ||
-          group == Group_Params::HYBRID_X25519_KYBER_512_R3_OQS ||
-          group == Group_Params::HYBRID_X25519_KYBER_768_R3_OQS ||
-          group == Group_Params::HYBRID_SECP256R1_KYBER_512_R3_OQS ||
-          group == Group_Params::HYBRID_SECP384R1_KYBER_768_R3_OQS ||
-          group == Group_Params::HYBRID_SECP521R1_KYBER_1024_R3_OQS;
-}
+      /**
+      * Returns std::nullopt if an unknown name
+      */
+      static std::optional<Group_Params> from_string(std::string_view group_name);
 
-constexpr bool is_kem(const Group_Params group) {
-   return is_pure_kyber(group) || is_hybrid(group);
-}
+      constexpr bool operator==(Group_Params_Code code) const { return m_code == code; }
 
-constexpr bool is_post_quantum(const Group_Params group) {
-   return is_pure_kyber(group) || is_hybrid(group);
-}
+      constexpr bool operator==(Group_Params other) const { return m_code == other.m_code; }
+
+      constexpr bool operator<(Group_Params other) const { return m_code < other.m_code; }
+
+      constexpr Group_Params_Code code() const { return m_code; }
+
+      constexpr uint16_t wire_code() const { return static_cast<uint16_t>(m_code); }
+
+      constexpr bool is_x25519() const { return m_code == Group_Params_Code::X25519; }
 
-std::string group_param_to_string(Group_Params group);
-Group_Params group_param_from_string(std::string_view group_name);
-std::vector<std::pair<std::string, std::string>> hybrid_group_param_to_algorithm_specs(Group_Params group);
-bool group_param_is_dh(Group_Params group);
+      constexpr bool is_ecdh_named_curve() const {
+         return m_code == Group_Params_Code::SECP256R1 || m_code == Group_Params_Code::SECP384R1 ||
+                m_code == Group_Params_Code::SECP521R1 || m_code == Group_Params_Code::BRAINPOOL256R1 ||
+                m_code == Group_Params_Code::BRAINPOOL384R1 || m_code == Group_Params_Code::BRAINPOOL512R1;
+      }
+
+      constexpr bool is_dh_named_group() const {
+         return m_code == Group_Params_Code::FFDHE_2048 || m_code == Group_Params_Code::FFDHE_3072 ||
+                m_code == Group_Params_Code::FFDHE_4096 || m_code == Group_Params_Code::FFDHE_6144 ||
+                m_code == Group_Params_Code::FFDHE_8192;
+      }
+
+      constexpr bool is_diffie_hellman() const {
+         const uint16_t group_id = wire_code();
+         return (group_id >= 256 && group_id < 512);
+      }
```

Comment:
😮 I'm sure that is some legacy definition in the allocation of this TLS identifier value space. Maybe add a comment at least?

Reply:
This is from https://www.rfc-editor.org/rfc/rfc7919 "Codepoints in the "Supported Groups Registry" with a high byte of 0x01 (that is, between 256 and 511, inclusive) are set aside for FFDHE groups" - that said it's unlikely we'll ever support other named DH groups and I don't think we'd ever actually call this function unless we specifically share the group with the peer.

Possibly something about this made more sense in TLS 1.0/1.1? I don't recall.

Reply:
I checked over all uses, everywhere we did use it we were implicitly assuming it was a named FFDHE group. So this predicate is removed entirely.